src/sysupdate/updatectl.c

```diff
@@ -0,0 +1,1226 @@
+        _cleanup_(version_clear) Version v = {};
+        char *version_json = NULL;
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *json = NULL;
+        int r;
+
+        assert(reply);
+        assert(ret);
+
+        r = sd_bus_message_read_basic(reply, 's', &version_json);
+
+        r = sd_json_parse(version_json, 0, &json, NULL, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse JSON: %m");
+
+        assert(sd_json_variant_is_object(json));
+
+        const char *key;
+        sd_json_variant *value;
+        JSON_VARIANT_OBJECT_FOREACH(key, value, json) {
+                if (streq(key, "version")) {
+                        assert(sd_json_variant_is_string(value));
+                        assert_se(v.version = strdup(sd_json_variant_string(value)));
+                } else if (streq(key, "newest")) {
+                        assert(sd_json_variant_is_boolean(value));
+                        if (sd_json_variant_boolean(value))
+                                v.flags |= UPDATE_NEWEST;
+                } else if (streq(key, "available")) {
+                        assert(sd_json_variant_is_boolean(value));
+                        if (sd_json_variant_boolean(value))
+                                v.flags |= UPDATE_AVAILABLE;
+                } else if (streq(key, "installed")) {
+                        assert(sd_json_variant_is_boolean(value));
+                        if (sd_json_variant_boolean(value))
+                                v.flags |= UPDATE_INSTALLED;
+                } else if (streq(key, "obsolete")) {
+                        assert(sd_json_variant_is_boolean(value));
+                        if (sd_json_variant_boolean(value))
+                                v.flags |= UPDATE_OBSOLETE;
+                } else if (streq(key, "protected")) {
+                        assert(sd_json_variant_is_boolean(value));
+                        if (sd_json_variant_boolean(value))
+                                v.flags |= UPDATE_PROTECTED;
+                } else if (streq(key, "changelog-url")) {
+                        sd_json_variant *entry;
+                        assert(sd_json_variant_is_array(value));
+                        JSON_VARIANT_ARRAY_FOREACH(entry, value) {
+                                assert(sd_json_variant_is_string(entry));
+                                r = strv_extend(&v.changelog, sd_json_variant_string(entry));
+                                if (r < 0)
+                                        log_error_errno(r, "Failed to parse JSON: %m");
+                        }
+                } else if (streq(key, "contents")) {
+                        r = sd_json_variant_format(value, 0, &v.contents_json);
+                        if (r < 0)
+                                log_error_errno(r, "Failed to parse JSON: %m");
+                } else
+                        assert_not_reached();
```

Comment:
please handle extensions gracefully.

and more importantly, use sd_json_dispatch() instead of this loop